pallets/contracts/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+    {
+        /// Emitted when instantiation fee of a template get changed.
+        /// IdentityId of the owner, Code hash of the template, old instantiation fee, new instantiation fee.
+        InstantiationFeeChanged(IdentityId, CodeHash, Balance, Balance),
+        /// Emitted when the instantiation of the template get frozen.
+        /// IdentityId of the owner, Code hash of the template.
+        InstantiationFreezed(IdentityId, CodeHash),
+        /// Emitted when the instantiation of the template gets un-frozen.
+        /// IdentityId of the owner, Code hash of the template.
+        InstantiationUnFreezed(IdentityId, CodeHash),
+    }
+}
+
+decl_module! {
+    // Wrap dispatchable functions for contracts so that we can add additional gating logic.
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+
+        /// Initialize the default event for this module.
+        fn deposit_event() = default;
+
+        /// Error type.
+        type Error = Error<T>;
+
+        /// The minimum amount required to generate a tombstone.
+        const NetworkShareInInstantiationFee: Perbill = T::NetworkShareInFee::get();
+
+        // Simply forwards to the `update_schedule` function in the Contract module.
+        #[weight = pallet_contracts::Call::<T>::update_schedule(schedule).get_dispatch_info().weight]
+        pub fn update_schedule(origin, schedule: Schedule) -> DispatchResult {
+            <pallet_contracts::Module<T>>::update_schedule(origin, schedule)
+        }
+
+        /// Simply forwards to the `put_code` function in the Contract module.
+        ///
+        /// # Additional functionality
+        /// 1. Allow origin to pass some meta-details related to template code.
+        /// 2. Charge protocol fee for deploying the template.
+        #[weight = 50_000_000.saturating_add(pallet_contracts::Call::<T>::put_code(code.clone()).get_dispatch_info().weight)]
+        pub fn put_code(
+            origin,
+            meta_info: SmartExtensionMetadata<BalanceOf<T>>,
+            code: Vec<u8>
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+
+            // Save metadata related to the SE template
+            // Generate the code_hash here as well because there is no way
+            // to read it directly from the upstream `pallet-contracts` module.
+            let code_hash = T::Hashing::hash(&code);
+
+            // Call underlying function
+            <pallet_contracts::Module<T>>::put_code(origin, code)?;
+
+            // Charge the protocol fee
+            T::ProtocolFee::charge_fee(ProtocolOp::ContractsPutCode)?;
+            <TemplateMetaDetails<T>>::insert(code_hash, TemplateMetadata {
+                meta_info: meta_info,
+                owner: sender,
+                frozen: false
+            });
+            Ok(())
+        }
+
+        // Simply forwards to the `call` function in the Contract module.
```

Comment:
Should we allow an owner to transfer ownership to a different DID?

Reply:
Looking at the code, we currently store ownership against an AccountId rather than DID. Is there a technical reason we need to do this (rather than storing the owners DID)?

Reply:
Do you mean template ownership?